jenkins/deploy-clio-server.sh

```diff
@@ -1,210 +1,256 @@
+  local -r previous_app_backup="${previous_app_dir}.old"
+
+  # Clean up any leftover state from previous failed deploys.
+  ${ssh_cmd} <<-EOF
+  sudo rm -rf ${new_app_staging_dir} ${previous_app_backup} &&
+  sudo mkdir ${new_app_staging_dir} &&
+  sudo chgrp ${ssh_user} ${new_app_staging_dir} &&
+  sudo chmod g+w ${new_app_staging_dir}
 EOF
 
-# copy configs to /app.new
-${SCPCMD} -r ${TMPDIR}/* ${SSH_USER}@${CLIO_HOST}:${NEW_APP_STAGING_DIR}
-
-# stop running the existing container
-stopcontainer
-
-# move /app.old to /app.old.old
-# move /app to /app.old
-# move /app.new to /app
-# create pos log directory if it doesn't already exist
-${SSHCMD} <<-EOF
-  ([ ! -d ${PREVIOUS_APP_DIR} ] || sudo mv ${PREVIOUS_APP_DIR} ${PREVIOUS_APP_BACKUP_DIR}) &&
-  ([ ! -d ${APP_DIR} ] || sudo mv ${APP_DIR} ${PREVIOUS_APP_DIR}) &&
-  sudo mv ${NEW_APP_STAGING_DIR} ${APP_DIR} &&
-  sudo mkdir -p ${HOST_LOG_DIR}/${POS_LOG_DIR}
+  # Copy rendered configs to the staging directory.
+  ${scp_cmd} -r ${tmp}/* ${ssh_user}@${clio_domain_name}:${new_app_staging_dir}
+
+  # Stop anything that's running.
+  stop_containers "${ssh_cmd}" ${tmp}
+
+  # Create backups, in case we need to roll back.
+  ${ssh_cmd} <<-EOF
+  ([ ! -d ${previous_app_dir} ] || sudo mv ${previous_app_dir} ${previous_app_backup}) &&
+  ([ ! -d ${app_dir} ] || sudo mv ${app_dir} ${previous_app_dir}) &&
+  sudo mv ${new_app_staging_dir} ${app_dir} &&
+  sudo mkdir -p ${log_init_path}
 EOF
 
-# start running the new container
-startcontainer
-
-STATUS=
-REPORTED_VERSION=
-
-# try 540 times at 20 second intervals (3 hours total)
-# TODO: if we change the server to bind its port before recovery and report a "recovering" status of some kind,
-# we can be smarter about how long we wait here before rolling back.
-attempts=1
-
-while [[ "$STATUS" != "OK" || "$REPORTED_VERSION" != "$DOCKER_TAG" ]]
-do
-  if [ ${attempts} -le 540 ]
-  then
-    sleep 20
-    echo "Attempt $attempts"
-    STATUS=$(curl -fs https://"${CLIO_HOST}/health" | jq -r .search)
-    REPORTED_VERSION=$(curl -fs https://"${CLIO_HOST}/version" | jq -r .version)
-    attempts=$((attempts + 1))
+  # Start up the new containers.
+  start_containers "${ssh_cmd}" ${app_dir}
+}
+
+poll_clio_health() {
+  local -r clio_domain_name=$1 docker_tag=$2
+
+  local status
+  local reported_version
+  local attempts=1
+
+  # Poll the Clio server every 20 seconds for 3 hours (540 times).
+  # TODO: The server could be smarter about reporting recovery vs. error, allowing us
+  # to shrink this polling time.
+  while [[ "$status" != "OK" || "$reported_version" != "$docker_tag" ]]; do
+    if [ ${attempts} -le 540 ]; then
+      sleep 20
+      echo "Attempt $attempts"
+      status=$(curl -fs "https://${clio_domain_name}/health" | jq -r .search)
+      reported_version=$(curl -fs "https://${clio_domain_name}/version" | jq -r .version)
+      attempts=$((attempts + 1))
+    else
+      return 1
+    fi
+  done
+
+  return 0
+}
+
```

Comment:
This script is often creating strings, modifying them, and then using the results as command lines.
OK, that's what everyone does ... but ...
It is usually (always?) better to manage command lines as argument vectors (bash arrays)
to avoid quote and eval problems. (Let me know if you're interested in learning more.)